### ISSUE TITLE
More modular context implementation (custom renderers)

### ISSUE THREAD

**Issue by redwoodfavorite** [Famous/engine#288]
We were in a bit of a crunch when writing Context and I think a lot of things are hardcoded.  I would really like to see this refactored so that custom renderers can register their own commands.

For example, currently we handle the GL_AMBIENT_LIGHT command as a part of the switch statement in compositor.

`Context.js`
```
case 'GL_AMBIENT_LIGHT':
     if (!this.WebGLRenderer) this.initWebGL();
     this.WebGLRenderer.setAmbientLightColor(
           path,
           commands[++localIterator],
           commands[++localIterator],
           commands[++localIterator]
      );
      break;
```

A more modular way to do this would be to have the renderer have a list of its commands to register.

`WebGLRenderer.js`
```
WebGLRenderer.COMMANDS = {
	'GL_AMBIENT_LIGHT': 'setAmbientLightColor'
}
```

And context would simply register renderers and all their commands.

`Context.js`
```
this.addRenderer('webgl-default', WebGLRenderer);
```

As far as handling how many arguments are passed in, I would propose that we just pass the path, command array, and local iterator into every handler function. For example:

`WebGLRenderer.js`
```
WebGLRenderer.prototype.setAmbientLightColor = function setAmbientLightColor(path, commands, iterator) {
    this.ambientLightColor[0] = commands[++iterator];
    this.ambientLightColor[1] = commands[++iterator];
    this.ambientLightColor[2] = commands[++iterator];
    return iterator;
};
```

I think this would make both adding new renderers and adding new commands easier.

This would also let us avoid having this everywhere in Context:

`Context.js`
```
if (this.webGLRenderer) {
   ...
}
```

P.S. I know https://github.com/Famous/engine/pull/281 is changing the switch statement but it shouldn't affect this implementation all that much.

**Comment by redwoodfavorite** [comment]
@adnan-wahab @DnMllr @michaelobriena @FarhadG @alexanderGugel

**Comment by michaelobriena** [comment]
I agree it is pretty messy.  I would like to table this until #281 gets merged

**Comment by alexanderGugel** [comment]
I like that! Especially `this.addRenderer('webgl-default', WebGLRenderer);`!

**Comment by adnan-wahab** [comment]
TABLED

**Comment by TheAlphaNerd** [comment]
ya'll aborbs

![bff](https://31.media.tumblr.com/tumblr_m3o7w0G1Og1qzx0km.gif)

**Comment by michaelobriena** [comment]
Will review this since the transform changes were merged

**Comment by alexanderGugel** [comment]
@michaelobriena I still think this is a good idea. Do you want to get this in?
